File: Course_2-Data_Structures/Week-3/Excercise_Challenges/3_merging_tables/merging_tables.cpp
#include <cstdio>
#include <cstdlib>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

class DisjointSets{
private:
	int n;
	int m;
	vector<int> list;
	vector<int> parent;
	vector<int> rank;
	int mx_size = 0;

	void write_data(void)
	{
		cout << mx_size << endl;
	}

	int getParent(int table)
	{
		vector<int> v;
		int root = table;

		while(root != parent[root]){
			v.push_back(parent[root]);
			root = parent[root];
		}

		for(auto a : v)
			parent[a] = root;

		return root;
	}

	void merge(int destination, int source)
	{
		int d_i = getParent(destination - 1);
		int s_i = getParent(source - 1);

		if(d_i == s_i)
			return;

		if(rank[s_i] >= rank[d_i])
			parent[s_i] = d_i;
		else{
			parent[d_i] = s_i;
			if(rank[s_i] == rank[d_i])
				rank[s_i] += 1;
		}

		list[d_i] += list[s_i];
		list[s_i] = 0;
		mx_size = max(mx_size, list[d_i]);
	}

	void read_data(void)
	{
		cin >> n >> m;

		parent.resize(n);
		for(int i = 0; i < n; ++i){
			parent[i] = i;
		}
		
		rank.resize(n, 0);

		list.resize(n);
		for(int i = 0; i < n; ++i){
			cin >> list[i];
			mx_size = max(mx_size, list[i]);
		}

		for(int i = 0, d, s; i < m; ++i){
			cin >> d >> s;

			merge(d, s);
			write_data();
		}
	}

public:
	void solve(void)
	{
		read_data();
	}
};

int main(void)
{
	ios_base::sync_with_stdio(false);
	DisjointSets disjoint;
	disjoint.solve();

	return 0;
}

File: Course_2-Data_Structures/Week-3/Excercise_Challenges/3_merging_tables/merging_tables.py
class DisjointSets:
    def __init__(self):
        self.n = 0
        self.m = 0
        self.data = []
        self.parent = []
        self.rank = []
        self.max_size = 0

    def get_parent(self, table):
        update_parent = []

        root = table
        while root != self.parent[root]:
            update_parent.append(self.parent[root])
            root = self.parent[root]

        for i in update_parent:
            self.parent[i] = root

        return root


    def merge(self, dst, src):
        src_parent = self.get_parent(src)
        dst_parent = self.get_parent(dst)

        if src_parent == dst_parent:
            return

        if self.rank[src_parent] >= self.rank[dst_parent]:
            self.parent[src_parent] = dst_parent
        else:
            self.parent[dst_parent] = src_parent
            if self.rank[src_parent] == self.rank[dst_parent]:
                self.rank[src_parent] += 1

        self.data[dst_parent] += self.data[src_parent]
        self.data[src_parent] = 0

        self.max_size = max(self.max_size, self.data[dst_parent])

    def write_data(self):
        print(self.max_size)

    def read_data(self):
        self.n, self.m = map(int, input().split())
        self.data = list(map(int, input().split()))

        self.parent = [i for i in range(self.n)]
        self.rank = [0] * self.n
        self.max_size = max(self.data)

        for _ in range(self.m):
            destination, source = map(int, input().split())

            self.merge(destination - 1, source - 1)
            self.write_data()


if __name__ == "__main__":
    obj = DisjointSets()
    obj.read_data()
